packages/container/libraries/strongly-typed/src/container.ts

```diff
@@ -0,0 +1,197 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-unnecessary-type-parameters */
+import type { interfaces } from 'inversify';
+
+type IfAny<T, TYes, TNo> = 0 extends 1 & T ? TYes : TNo;
+
+type BindingMapProperty = string | symbol;
+type BindingMap = Record<BindingMapProperty, any>;
+type BindingMapKey<T extends BindingMap> = keyof T & BindingMapProperty;
+type MappedServiceIdentifier<T extends BindingMap> = IfAny<
+  T,
+  interfaces.ServiceIdentifier,
+  BindingMapKey<T>
+>;
+type ContainerBinding<
+  T extends BindingMap,
+  TKey extends MappedServiceIdentifier<T> = any,
+> = TKey extends keyof T
+  ? T[TKey]
+  : TKey extends interfaces.Newable<infer C>
+    ? C
+    : TKey extends interfaces.Abstract<infer D>
+      ? D
+      : never;
+
+type First<T extends any[]> = T extends [infer TFirst, ...any[]]
+  ? TFirst
+  : never;
+
+type AllButFirst<T extends any[]> = T extends [any, ...infer TRest]
+  ? TRest
+  : never;
+
+interface ContainerOverrides<
+  T extends BindingMap = any,
```

Comment:
Now that these type definitions are opt-in, we *could* potentially make this a **required** type definition (otherwise why are you even using this interface), which might simplify types?

On the other hand, I quite like that this is vaguely backwards-compatible still, and possibly helps if you want to allow weird hierarchies involving children/parents:

```ts
const parent = new Container() as TypedContainer<ParentMap>;
const child = parent.createChild() // can create without binding map
// ...but this is still strongly typed despite `child` not having a mapping
child.parent.get('...') 
```

Reply:
As discussed, we'll keep this as optional:

- makes it easier to move into `core` later if we want to
- probably works nicer with hierarchies
- more permissive, which is generally a bit nicer for consumers